File: solutions/scala/wordy/1/src/main/scala/Wordy.scala
object Wordy {
  def answer(question: String): Option[Int] =
    val expression = question
      .stripPrefix("What is ")
      .stripSuffix("?")
      .replace("multiplied by", "multiplied")
      .replace("divided by", "divided")
      .split(" ")
      .toList

    expression match {
      case first :: rest => first.toIntOption.flatMap(evaluate(_, rest))
      case _ => None
    }

  private def evaluate(current: Int, tokens: List[String]): Option[Int] =
    tokens match {
      case Nil => Some(current)
      case operator :: value :: rest =>
        for
          number <- value.toIntOption
          result <- applyOperator(current, operator, number)
          answer <- evaluate(result, rest)
        yield answer
      case _ => None
    }

  private def applyOperator(left: Int, operator: String, right: Int): Option[Int] =
    operator match {
      case "plus" => Some(left + right)
      case "minus" => Some(left - right)
      case "multiplied" => Some(left * right)
      case "divided" => Some(left / right)
      case _ => None
    }
}

File: solutions/kotlin/wordy/1/src/main/kotlin/Wordy.kt
object Wordy {
    fun answer(input: String): Int {
        val tokens =
            input
                .removePrefix("What is ")
                .removeSuffix("?")
                .replace("multiplied by", "multiplied")
                .replace("divided by", "divided")
                .replace(POWER_PATTERN, "raised $1")
                .split(" ")

        val first = tokens.firstOrNull()?.toIntOrNull() ?: throw invalidQuestion()

        return evaluate(first, tokens.drop(1))
    }

    private fun evaluate(current: Int, tokens: List<String>): Int = when {
        tokens.isEmpty() -> current
        tokens.size < 2 -> throw invalidQuestion()
        else -> {
            val operator = tokens[0]
            val operand = tokens[1].toIntOrNull() ?: throw invalidQuestion()
            val result = applyOperator(current, operator, operand)

            evaluate(result, tokens.drop(2))
        }
    }

    private fun applyOperator(left: Int, operator: String, right: Int): Int = when (operator) {
        "plus" -> left + right
        "minus" -> left - right
        "multiplied" -> left * right
        "divided" -> left / right
        "raised" -> left.power(right)
        else -> throw invalidQuestion()
    }

    private fun Int.power(exponent: Int): Int = (1..exponent).fold(1) { result, _ -> result * this }

    private fun invalidQuestion() = IllegalArgumentException("Invalid question")

    private val POWER_PATTERN = Regex("raised to the (\\d+)(?:st|nd|rd|th) power")
}
